src/runtime/auto-prompt-manager.ts

```diff
@@ -1134,51 +1172,131 @@ export class AutoPromptManager {
   }
 
   /**
-   * Fetches dismissal timestamps from local storage for a given action type.
+   * Fetches impression timestamps from local storage for a given action type.
    */
-  private async getActionDismissals_(actionType: string): Promise<number[]> {
-    if (!ACTION_TO_DISMISSAL_STORAGE_KEY_MAP.has(actionType)) {
-      // TODO(justinchou): Update to new error event for dismissals.
+  private async getActionImpressions_(actionType: string): Promise<number[]> {
+    if (!ACTION_TO_IMPRESSION_STORAGE_KEY_MAP.has(actionType)) {
       this.eventManager_.logSwgEvent(
         AnalyticsEvent.EVENT_ACTION_IMPRESSIONS_STORAGE_KEY_NOT_FOUND_ERROR
       );
       return [];
     }
 
     return this.storage_.getFrequencyCappingEvent(
-      ACTION_TO_DISMISSAL_STORAGE_KEY_MAP.get(actionType)!
+      ACTION_TO_IMPRESSION_STORAGE_KEY_MAP.get(actionType)!
     );
   }
 
-  /**
-   * Fetches impression timestamps from local storage for a given action type.
-   */
-  private async getActionImpressions_(actionType: string): Promise<number[]> {
-    if (!ACTION_TO_IMPRESSION_STORAGE_KEY_MAP.has(actionType)) {
-      this.eventManager_.logSwgEvent(
-        AnalyticsEvent.EVENT_ACTION_IMPRESSIONS_STORAGE_KEY_NOT_FOUND_ERROR
+  async getTimestamps(): Promise<LocalStorageTimestamps[]> {
+    const stringified = await this.storage_.get(
+      StorageKeys.TIMESTAMPS,
+      /* useLocalStorage */ true
+    );
+    const array = stringified ? JSON.parse(stringified) : [];
+
+    return array
+      .map((t: LocalStorageTimestamps) => {
+        const impressions = t.impressions
+          ? pruneTimestamps(t.impressions, TWO_WEEKS_IN_MILLIS)
+          : [];
+        const dismissals = t.dismissals
+          ? pruneTimestamps(t.dismissals, TWO_WEEKS_IN_MILLIS)
+          : [];
+        const completions = t.completions
+          ? pruneTimestamps(t.completions, TWO_WEEKS_IN_MILLIS)
+          : [];
+        return {
+          action: t.action,
+          ...(impressions?.length && {impressions}),
```

Comment:
Can we just set empty array instead of leaving it undefined? I think making timestamp array fields in `LocalStorageTimestamps` interface required would be more strongly typed and easier to process since we don't need to check the presence. WDYT?

With the suggested change above, it can be

```
return {
  action: t.action,
  impressions: pruneTimestamps(t.impressions || [], TWO_WEEKS_IN_MILLIS),
  ...
}
```

Reply:
The tradeoff here is optimizing for timestamps lookups / fetching vs. minimizing local storage bloating. 

I think changing `LocalStorageTimestamps` to a map by action type makes sense, with that in mind we may have the following local storage timestamps:
`
{
    'TYPE_CONTRIBUTION': {impressions: [], dismissals: [], completions: []},
    'TYPE_REWARDED_SURVEY: {impressions: [], dismissals: [], completions: []},
    'TYPE_REGISTRATION_WALL: {impressions: [], dismissals: [], completions: []},
}
`

One possible benefit that I can think of is that this type of local storage can be used to infer that a reader at least had a prompt impression > 2 weeks ago. Otherwise, things that are written to local storage will always persist (other than pruned timestamps). WDYT?

Reply:
Created https://github.com/subscriptions-project/swg-js/pull/3456.